fix for LDAPServerURIField error if number present in top-level-domain

fosterseth opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
related to issue #3646
continuation of this closed PR https://github.com/ansible/awx/pull/4774

LDAP server URI field does not validate if top-level-domain contains a number, throwing a generic 400 error to the user.

The URI field is passed to a built-in django validator that does a regex match to check for validity.

This URI validator can take in a custom regex. My solution sets up a custom regex in the LDAPServerURIField class and passes it to the URLValidator that is appended in the parent class init.

For the custom regex, I simply grabbed the default URLValidator regex defined here,
https://github.com/django/django/blob/55b68de643b5c2d5f0a8ea7587ab3b2966021ccc/django/core/validators.py#L89

and added number capabilities,

`r'(?:[a-z' + URLValidator.ul + r'0-9' + '-]{2,63}' # domain label`

Added a unit test to make sure valid URIs pass, and invalid URIs catch the correct exception.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 7.0.0

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

AlanCoding commented on awx/sso/tests/unit/test_fields.py:
```diff
@@ -114,3 +115,23 @@ def test_internal_value_invalid(self, group_type, data, expected):
+
+    invalid_URIs = [
+        r'ldaps://servername.-so3:444'
+    ]
```
Could you use the pytest parameterization for these? like:

https://github.com/ansible/awx/blob/d9965cfe7e0f7a82fea04acb3b6b49852874f655/awx/main/tests/unit/utils/test_filters.py#L15

that helps to make it clear which case the test failed on if it fails.

fosterseth replied:
thanks for the suggestion, I changed the tests to include pytest.mark.parametrize()

AlanCoding commented on awx/sso/fields.py:
```diff
@@ -233,12 +234,34 @@ def _default_from_required_settings(self):
+    tld_re = (
+        r'\.'                                              # dot
+        r'(?!-)'                                           # can't start with a dash
+        r'(?:[a-z' + URLValidator.ul + r'0-9' + '-]{2,63}' # domain label
```
Is this what changed from the original Django class? It might be good to point that out in a comment right here.

fosterseth replied:
I made a comment to indicate which line was changed from django URLValidator

jbradberry reviewed (commented):
I concur with Alan's review.

jbradberry reviewed (approved):
Seems good to me.

fosterseth committed:
Fix LDAPServerURIField number in domain

- Bug: API error if LDAPServerURIField contains a number in the top level domain
- Add custom regex in LDAPServerURIField class that is passed to django
  URLValidator
- The custom regex allows for numbers to be present in the top level domain
- Unit tests check that valid URIs pass through URLValidator, and that
  invalid URIs raise the correct exception
- Related to issue #3646

ryanpetrello commented on awx/sso/fields.py:
```diff
@@ -233,12 +234,34 @@ def _default_from_required_settings(self):
+    tld_re = (
+        r'\.'                                              # dot
+        r'(?!-)'                                           # can't start with a dash
+        r'(?:[a-z' + URLValidator.ul + r'0-9' + '-]{2,63}' # domain label, this line was changed from the original URLValidator
```
I think we should try contributing this back upstream to Django itself.
